src/bson.c

```diff
@@ -666,23 +870,30 @@ static bool php_phongo_bson_visit_document(const bson_iter_t* iter ARG_UNUSED, c
 			 * visitor. Free state.zchild, which we just initialized, and return
 			 * true to stop iteration for our parent context. */
 			zval_ptr_dtor(&state.zchild);
+			php_phongo_bson_state_dtor(&state);
 			return true;
 		}
+
+		php_phongo_bson_state_dtor(&state);
+		php_phongo_field_path_pop(parent_state->field_path);
 	}
 
 	return false;
 } /* }}} */
 
 static bool php_phongo_bson_visit_array(const bson_iter_t* iter ARG_UNUSED, const char* key, const bson_t* v_array, void* data) /* {{{ */
 {
-	zval*       retval = PHONGO_BSON_STATE_ZCHILD(data);
-	bson_iter_t child;
+	zval*                  retval = PHONGO_BSON_STATE_ZCHILD(data);
+	bson_iter_t            child;
+	php_phongo_bson_state* parent_state = (php_phongo_bson_state*) data;
 	TSRMLS_FETCH();
 
+	php_phongo_field_path_push(((php_phongo_bson_state*) data)->field_path, key);
```

Comment:
`((php_phongo_bson_state*) data)` can just be `parent_state`, as you did in `php_phongo_bson_visit_document()`.